src/Idris/REPLCommon.idr

```diff
@@ -51,32 +51,39 @@ printError : {auto o : Ref ROpts REPLOpts} ->
              Doc IdrisAnn -> Core ()
 printError msg = printWithStatus (pretty "error") msg
 
--- Display an error message from checking a source file
+DocCreator : Type -> Type
+DocCreator a = a -> Core (Doc IdrisAnn)
+
 export
-emitError : {auto c : Ref Ctxt Defs} ->
+emitProblem : {auto c : Ref Ctxt Defs} ->
             {auto o : Ref ROpts REPLOpts} ->
             {auto s : Ref Syn SyntaxInfo} ->
-            Error -> Core ()
-emitError err
+            a -> (DocCreator a) -> (DocCreator a) -> (a -> Maybe FC) -> Core ()
+emitProblem a replDocCreator idemodeDocCreator getFC
     = do opts <- get ROpts
          case idemode opts of
               REPL _ =>
-                  do msg <- display err >>= render
+                  do msg <- replDocCreator a >>= render
                      coreLift $ putStrLn msg
               IDEMode i _ f =>
-                  do msg <- perror err
-                     case getErrorLoc err of
+                  do msg <- idemodeDocCreator a
+                     case map defaultFC (getFC a) of
                           Nothing => iputStrLn msg
-                          Just fc =>
+                          Just (file, startPos, endPos) =>
                             send f (SExpList [SymbolAtom "warning",
-                                   SExpList [toSExp (file fc),
-                                            toSExp (addOne (startPos fc)),
-                                              toSExp (addOne (endPos fc)),
+                                   SExpList [toSExp file,
+                                            toSExp (addOne startPos),
+                                              toSExp (addOne endPos),
                                               toSExp !(renderWithoutColor msg),
                                               -- highlighting; currently blank
                                               SExpList []],
                                     toSExp i])
   where
+    -- TODO: Display a better message when the error doesn't contain a location
+    defaultFC : FC -> NonEmptyFC
+    defaultFC EmptyFC = ("", (0, 0), (0, 0))
+    defaultFC (MkFC fname start end) = (fname, start, end)
+
```

Comment:
This should be moved to `Core.FC`. It could be reused in other places where you're doing
essentially this (e.g. the definition of `constructSExp`).